### PR TITLE
Adapt template to new architecture autolinking on Android

### PR DESCRIPTION
**Pull request by thymikee**
## Summary

Provides necessary changes for the autolinking to work in new architecture on Android. Depends on https://github.com/react-native-community/cli/pull/1603 and is subject to change. 

Upgraded the RN CLI to v9.0.0-alpha.0 so that it's testable locally.

## Changelog

[Android] [Changed] - Adapt template to new architecture autolinking on Android

## Test Plan

CI

### CODE REVIEW

File: template/android/app/src/main/jni/Android.mk [hunk 1]
```diff
@@ -5,6 +5,10 @@ include $(REACT_ANDROID_DIR)/Android-prebuilt.mk
 # If you wish to add a custom TurboModule or Fabric component in your app you
 # will have to include the following autogenerated makefile.
 # include $(GENERATED_SRC_DIR)/codegen/jni/Android.mk
+
+# Includes the MK file for autolinked libraries
++include $(PROJECT_BUILD_DIR)/generated/rncli/src/main/jni/Android-rncli.mk
```
Comment (cortinico) [inline template/android/app/src/main/jni/Android.mk:10]:
```suggestion
include $(PROJECT_BUILD_DIR)/generated/rncli/src/main/jni/Android-rncli.mk  
```

File: template/android/app/src/main/jni/MainComponentsRegistry.cpp [hunk 2]
```diff
@@ -14,6 +15,9 @@ std::shared_ptr<ComponentDescriptorProviderRegistry const>
 MainComponentsRegistry::sharedProviderRegistry() {
   auto providerRegistry = CoreComponentsRegistry::sharedProviderRegistry();
 
+  // Providers registered automatically by RN CLI
```
Comment (cortinico) [inline template/android/app/src/main/jni/MainComponentsRegistry.cpp:18]: Please mention autolinking in this comment here

Comment (thymikee) [inline template/android/app/src/main/jni/MainComponentsRegistry.cpp:18]:
```suggestion
  // Autolinked providers registered by RN CLI
```

**Comment by cortinico** [comment]
> 



> [react-native-community/cli#1603](https://github.com/react-native-community/cli/pull/1603) is now merged and released in 9.0.0-alpha.0, so this PR is now unblocked and ready to be merged.

Awesome, thanks for the heads up 👍 
I'll import this and get back to you with some testing.

**Comment by cortinico** [comment]
This is currently blocked by https://github.com/facebook/react-native/pull/33859
which needs to land before this (and I'm having some difficulty merging internally). Will get back to this

**Comment by thymikee** [comment]
@cortinico looks like https://github.com/facebook/react-native/pull/33859 is merged

**Comment by cortinico** [comment]
> @cortinico looks like #33859 is merged

Yup 👍 

Don't you mind rebase this one? 
CI should be green now + I've managed to let this build internally

**Comment by thymikee** [comment]
Rebased :)

**Comment by cortinico** [comment]
Hey @thymikee Sorry for the delay. We're going to merge this as it is as I was able to let it work correctly for the majority of cases. 

However there are a couple of changes we'd like to address on the CLI side of things:

- Autolinking is currently breaking libraries if they don't define a `libraryName = ` field. The problem is here: https://github.com/react-native-community/cli/blob/d5ab2143eeec5d250e6cf13007156f0b042e2c57/packages/platform-android/native_modules.gradle#L259-L264 - Effectively we're importing the `Android.mk` file regardless of a library actually using codegen or not. This means that if you autolink a library that doesn't have codegen set up (i.e. `libraryName = ` is not set), you'll fail the build with an import of a missign file.
- The whole reliance on `libraryName = ` needs to be adapted a bit. 
  - Specifically here: https://github.com/react-native-community/cli/blob/d5ab2143eeec5d250e6cf13007156f0b042e2c57/packages/platform-android/src/config/findLibraryName.ts#L8 - if you specify `libraryName = 'abc'` instead of `libraryName = "abc"` (single quotes) which is valid in Gradle, it won't work.
  - Technically libraries could omit the `libraryName =` field entierely and use the default. This is not handled (though they could override the autolinking config).
  - The whole setup needs to be adapted to read the `codegenConfig` from `package.json` and fallback to `build.gradle` if needed.
  - If the user is using `build.gradle.kts` -> Autolinking won't work
  - If the user is usign any other Gradle plugin which has a `libraryName` field, this mechanism will break
- CMake support is missing.

I have PRs for some of those points that I will be sending towards the cli repo in the next days 👍

**Pull request closed without merging**